plugins/dell-dock2/fu-dell-dock2-ec.c

```diff
@@ -0,0 +1,972 @@
+	guint8 cmd = DELL_DOCK2_EC_HID_CMD_GET_DOCK_TYPE;
+
+	/* expect response 1 byte */
+	if (!fu_dell_dock2_ec_read(device, cmd, res, error)) {
+		g_prefix_error(error, "Failed to query dock type: ");
+		return FALSE;
+	}
+
+	self->base_type = res->data[0];
+
+	/* check dock type to proceed with this plugin or exit as unsupported */
+	return fu_dell_dock2_ec_dock_type_extract(device, error);
+}
+
+static gboolean
+fu_dell_dock2_ec_dock_info_extract(FuDevice *device, GError **error)
+{
+	FuDellDock2Ec *self = FU_DELL_DOCK2_EC(device);
+
+	if (!self->dock_info) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "Failed to parse dock info");
+		return FALSE;
+	}
+
+	if (self->dock_info->header.total_devices == 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_SIGNATURE_INVALID,
+			    "No bridge devices detected, dock may be booting up");
+		return FALSE;
+	}
+	g_info("found %u devices [%u->%u]",
+	       self->dock_info->header.total_devices,
+	       self->dock_info->header.first_index,
+	       self->dock_info->header.last_index);
+
+	for (guint i = 0; i < self->dock_info->header.total_devices; i++) {
+		struct FuDellDock2EcQueryEntry dev_entry = self->dock_info->devices[i];
+		const gchar *type_str;
+		const gchar *location_str;
+		guint32 version32 = GUINT32_FROM_BE(dev_entry.version.version_32);
+
+		/* name the component */
+		type_str = fu_dell_dock2_ec_devicetype_to_str(dev_entry.ec_addr_map.device_type,
+							      dev_entry.ec_addr_map.sub_type,
+							      dev_entry.ec_addr_map.instance);
+		if (type_str == NULL)
+			continue;
+
+		/* name the location of component */
+		location_str = (dev_entry.ec_addr_map.location == DELL_DOCK2_EC_LOCATION_BASE)
+				   ? "Base"
+				   : "Module";
+
+		/* show the component location */
+		g_debug("#%u: %s located in %s (A: %u I: %u)",
+			i,
+			type_str,
+			location_str,
+			dev_entry.ec_addr_map.arg,
+			dev_entry.ec_addr_map.instance);
+
+		/* show the component version */
+		g_debug("\tVersion32: %08x, Version8: %s",
+			dev_entry.version.version_32,
+			fu_hex_version_from_uint32(version32, FWUPD_VERSION_FORMAT_QUAD));
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_ec_dock_info_cmd(FuDevice *device, GError **error)
+{
+	FuDellDock2Ec *self = FU_DELL_DOCK2_EC(device);
+	gsize length = sizeof(FuDellDock2DockInfoStructure);
+	g_autoptr(GByteArray) res = g_byte_array_new_take(g_malloc0(length), length);
+	guint8 cmd = DELL_DOCK2_EC_HID_CMD_GET_DOCK_INFO;
+
+	/* get dock info over HID */
+	if (!fu_dell_dock2_ec_read(device, cmd, res, error)) {
+		g_prefix_error(error, "Failed to query dock info: ");
+		return FALSE;
+	}
+	if (res->len != length) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "invalid dock info size: expected %" G_GSIZE_FORMAT ",got %u",
+			    length,
+			    res->len);
+		return FALSE;
+	}
+	memcpy(self->dock_info, res->data, length);
```

Comment:
I think you should be using `fu_memcpy_safe` here.

Reply:
Fixed, thanks.